File: ios/LaneShadow/Views/Atoms/CompassPlusIcon.swift
import LaneShadowTheme
import SwiftUI

// MARK: - CompassPlusIcon Component

/**
 * CompassPlusIcon component
 *
 * SVG compass icon with plus badge at bottom-right
 *
 * ## Design Tokens Used
 * - Colors:
 *   - `theme.colors.primary.default` - Compass circle fill
 *   - `theme.colors.onPrimary.default` - Compass needle and N/S marks
 *   - `theme.colors.onSurface.default` - Plus badge background
 *   - `theme.colors.surface.default` - Plus sign foreground
 * - Space:
 *   - `theme.space.xs` (4) - Used for stroke width calculation
 *   - `theme.space.md` (12) - Used for badge radius calculation
 *
 * ## Parameters
 * - Parameters:
 *   - size: Overall icon size (default: 48)
 *   - label: Optional accessibility label (default: "Compass with plus badge")
 *   - testID: Test identifier for UI testing
 */
public struct CompassPlusIcon: View {
    @Environment(\.theme) private var theme

    private let size: CGFloat
    private let label: String?
    private let testID: String?

    public init(
        size: CGFloat = 48,
        label: String? = nil,
        testID: String? = nil
    ) {
        self.size = size
        self.label = label
        self.testID = testID
    }

    // MARK: - Computed Properties

    private var strokeWidth: CGFloat {
        max(1.5, theme.space.xs / 3)
    }

    private var badgeRadius: CGFloat {
        max(6, theme.space.md) / 2
    }

    private var center: CGPoint {
        CGPoint(x: size / 2, y: size / 2)
    }

    private var radius: CGFloat {
        (size - strokeWidth) / 2
    }

    private var badgeCenter: CGPoint {
        CGPoint(x: size - badgeRadius, y: size - badgeRadius)
    }

    // MARK: - Body

    public var body: some View {
        Canvas { context, size in
            drawCompass(in: context, size: size)
            drawBadge(in: context, size: size)
        }
        .frame(width: size, height: size)
        .accessibilityElement(children: .ignore)
        .accessibilityLabel(label ?? "Compass with plus badge")
        .accessibilityIdentifier(testID ?? "compass-plus-icon")
    }

    // MARK: - Drawing Methods

    private func drawCompass(in context: GraphicsContext, size: CGSize) {
        // Draw compass circle
        context.fill(
            Path(ellipseIn: CGRect(center: center, radius: radius)),
            with: .color(theme.colors.primary.default)
        )

        // Draw compass circle stroke
        context.stroke(
            Path(ellipseIn: CGRect(center: center, radius: radius)),
            with: .color(theme.colors.onPrimary.default),
            lineWidth: strokeWidth
        )

        // Draw diamond needle
        drawNeedle(in: context)

        // Draw North/South marks
        drawMarks(in: context)
    }

    private func drawNeedle(in context: GraphicsContext) {
        let needleWidth = radius * 0.25
        let needleHeight = radius * 0.6

        var needlePath = Path()
        needlePath.move(to: CGPoint(x: center.x, y: center.y - needleHeight / 2))
        needlePath.addLine(to: CGPoint(x: center.x + needleWidth / 2, y: center.y))
        needlePath.addLine(to: CGPoint(x: center.x, y: center.y + needleHeight / 2))
        needlePath.addLine(to: CGPoint(x: center.x - needleWidth / 2, y: center.y))
        needlePath.closeSubpath()

        context.fill(needlePath, with: .color(theme.colors.onPrimary.default))
    }

    private func drawMarks(in context: GraphicsContext) {
        let markLength = radius * 0.15
        let markInset = radius * 0.1

        // North mark
        var northPath = Path()
        northPath.move(to: CGPoint(x: center.x, y: center.y - radius + markInset))
        northPath.addLine(to: CGPoint(x: center.x, y: center.y - radius + markInset + markLength))

        // South mark
        var southPath = Path()
        southPath.move(to: CGPoint(x: center.x, y: center.y + radius - markInset))
        southPath.addLine(to: CGPoint(x: center.x, y: center.y + radius - markInset - markLength))

        context.stroke(northPath, with: .color(theme.colors.onPrimary.default), lineWidth: strokeWidth)
        context.stroke(southPath, with: .color(theme.colors.onPrimary.default), lineWidth: strokeWidth)
    }

    private func drawBadge(in context: GraphicsContext, size: CGSize) {
        // Draw badge circle
        context.fill(
            Path(ellipseIn: CGRect(center: badgeCenter, radius: badgeRadius)),
            with: .color(theme.colors.onSurface.default)
        )

        // Draw plus sign
        drawPlus(in: context)
    }

    private func drawPlus(in context: GraphicsContext) {
        let plusWidth = badgeRadius * 0.6
        let plusThickness = max(1.5, strokeWidth)

        // Horizontal line
        var horizontalPath = Path()
        horizontalPath.move(to: CGPoint(x: badgeCenter.x - plusWidth / 2, y: badgeCenter.y))
        horizontalPath.addLine(to: CGPoint(x: badgeCenter.x + plusWidth / 2, y: badgeCenter.y))

        // Vertical line
        var verticalPath = Path()
        verticalPath.move(to: CGPoint(x: badgeCenter.x, y: badgeCenter.y - plusWidth / 2))
        verticalPath.addLine(to: CGPoint(x: badgeCenter.x, y: badgeCenter.y + plusWidth / 2))

        context.stroke(horizontalPath, with: .color(theme.colors.surface.default), lineWidth: plusThickness)
        context.stroke(verticalPath, with: .color(theme.colors.surface.default), lineWidth: plusThickness)
    }
}

// MARK: - CGRect Extension for Center

private extension CGRect {
    init(center: CGPoint, radius: CGFloat) {
        self.init(
            x: center.x - radius,
            y: center.y - radius,
            width: radius * 2,
            height: radius * 2
        )
    }
}

// MARK: - Preview

#Preview("Default Size") {
    CompassPlusIcon()
        .laneShadowTheme()
}

#Preview("Small") {
    CompassPlusIcon(size: 32)
        .laneShadowTheme()
}

#Preview("Large") {
    CompassPlusIcon(size: 64)
        .laneShadowTheme()
}

#Preview("With Accessibility Label") {
    CompassPlusIcon(
        size: 48,
        label: "Add new route",
        testID: "add-route-icon"
    )
    .laneShadowTheme()
}

File: android/app/src/main/java/com/laneshadow/ui/components/atoms/CompassPlusIcon.kt
package com.laneshadow.ui.components.atoms

import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.size
import androidx.compose.runtime.Composable
import androidx.compose.ui.Modifier
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.Path
import androidx.compose.ui.graphics.drawscope.DrawScope
import androidx.compose.ui.graphics.drawscope.Stroke
import androidx.compose.ui.graphics.nativeCanvas
import androidx.compose.ui.platform.LocalDensity
import androidx.compose.ui.semantics.Role
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.role
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.semantics.testTag
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp
import com.laneshadow.theme.LocalLaneShadowTheme
import kotlin.math.max
import kotlin.math.min

/**
 * CompassPlusIcon - SVG compass icon with plus badge
 *
 * Renders a compass with:
 * - Filled circle with primary color and onPrimary stroke
 * - Diamond needle pointing up in onPrimary color
 * - North/South marks as vertical lines
 * - Plus badge in bottom-right quadrant
 *
 * @param size Icon size in density-independent pixels (default: 28dp)
 * @param label Accessibility label for screen readers
 * @param testID Test ID for UI testing
 * @param modifier Modifier for the icon container
 */
@Composable
fun CompassPlusIcon(
    size: Dp = 28.dp,
    label: String? = null,
    testID: String? = null,
    modifier: Modifier = Modifier,
) {
    val theme = LocalLaneShadowTheme.current

    Box(
        modifier = modifier
            .semantics {
                this.role = Role.Image
                label?.let {
                    contentDescription = it
                }
                testID?.let {
                    testTag = it
                }
            }
    ) {
        val density = LocalDensity.current
        val sizePx = with(density) { size.toPx() }

        androidx.compose.foundation.Canvas(
            modifier = Modifier.size(size),
            onDraw = {
                drawCompassPlusIcon(
                    size = sizePx,
                    primaryColor = theme.colors.primary.default,
                    onPrimaryColor = theme.colors.onPrimary.default,
                    onSurfaceColor = theme.colors.onSurface.default,
                    surfaceColor = theme.colors.surface.default,
                    strokeWidth = max(1.5f, theme.space.xs.toPx() / 3f),
                    badgeRadius = max(6f, theme.space.md.toPx()) / 2f,
                )
            },
        )
    }
}

/**
 * Draw the compass plus icon on the canvas
 */
private fun DrawScope.drawCompassPlusIcon(
    size: Float,
    primaryColor: Color,
    onPrimaryColor: Color,
    onSurfaceColor: Color,
    surfaceColor: Color,
    strokeWidth: Float,
    badgeRadius: Float,
) {
    val center = size / 2f
    val radius = (size - strokeWidth * 2) / 2f

    // Draw compass circle (filled with primary, stroked with onPrimary)
    drawCircle(
        color = primaryColor,
        radius = radius,
        center = Offset(center, center),
        style = Stroke(width = strokeWidth),
    )

    // Draw diamond needle (pointing up)
    val needlePath = Path().apply {
        val needleHeight = radius * 0.7f
        val needleWidth = radius * 0.4f

        // Diamond shape pointing up
        moveTo(center, center - needleHeight) // Top point
        lineTo(center + needleWidth, center) // Right point
        lineTo(center, center + needleHeight * 0.5f) // Bottom point (shorter)
        lineTo(center - needleWidth, center) // Left point
        close()
    }

    drawPath(
        path = needlePath,
        color = onPrimaryColor,
    )

    // Draw north/south marks (vertical lines)
    val markHeight = radius * 0.15f
    val markWidth = radius * 0.08f
    val markDistance = radius * 0.6f

    // North mark
    drawLine(
        color = onPrimaryColor,
        start = Offset(center, center - markDistance),
        end = Offset(center, center - markDistance + markHeight),
        strokeWidth = markWidth,
    )

    // South mark
    drawLine(
        color = onPrimaryColor,
        start = Offset(center, center + markDistance - markHeight),
        end = Offset(center, center + markDistance),
        strokeWidth = markWidth,
    )

    // Draw plus badge in bottom-right quadrant
    val badgeCenterX = center + radius * 0.5f
    val badgeCenterY = center + radius * 0.5f

    // Badge circle (filled with onSurface)
    drawCircle(
        color = onSurfaceColor,
        radius = badgeRadius,
        center = Offset(badgeCenterX, badgeCenterY),
    )

    // Plus sign (in surface color)
    val plusSize = badgeRadius * 0.6f
    val plusStroke = badgeRadius * 0.25f

    // Horizontal line of plus
    drawLine(
        color = surfaceColor,
        start = Offset(badgeCenterX - plusSize, badgeCenterY),
        end = Offset(badgeCenterX + plusSize, badgeCenterY),
        strokeWidth = plusStroke,
    )

    // Vertical line of plus
    drawLine(
        color = surfaceColor,
        start = Offset(badgeCenterX, badgeCenterY - plusSize),
        end = Offset(badgeCenterX, badgeCenterY + plusSize),
        strokeWidth = plusStroke,
    )
}
